### DISCUSSION TITLE
Is moby affected by vulnerabilities in the vendored buildkit?

### DISCUSSION THREAD

**Discussion by dcermak** [moby/moby#47383] [Q&A] [answered]
Moby/Docker version 24 still vendors in buildkit v0.11 which has multiple vulnerabilities. Are these actually affecting moby itself, if the user has a non-vulnerable version of buildx installed or not installed buildx at all? I'm wondering whether the code paths of buildkit are even reachable from moby with buildx not installed.

**Comment by thaJeztah** [comment, marked as answer]
Yes `docker build` with the built-in BuildKit builder is still affected by the CVEs in BuildKit; see https://github.com/moby/moby/discussions/47328#discussioncomment-8369700

**Reply by thaJeztah** [reply]
I saw there were contributions to backport the fixes to the v0.11 and v0.10 branches though (but not yet merged / accepted by the BuildKit maintainers);

- https://github.com/moby/buildkit/pull/4638
- https://github.com/moby/buildkit/pull/4639
- https://github.com/moby/buildkit/pull/4640
- https://github.com/moby/buildkit/pull/4642

**Reply by dcermak** [reply]
Thank you for the confirmation! Just out of interest, how would one invoke the built-in buildkit builder? I would like to make sure that my fixes are actually working.

**Reply by thaJeztah** [reply]
> how would one invoke the built-in buildkit builder?

BuildKit is the default builder used since Docker 23.0, which means that any `docker build`, `docker image build` and `docker buildx build` command is by default handled by the built-in BuildKit builder. There's some exceptions to that;

- when executing the `docker build` command with `DOCKER_BUILDKIT=0`, it uses the classic / legacy builder
- if the `buildx` CLI component (plugin) is not installed, it falls back to the classic / legacy builder
- (third-party) clients executing builds through the `/build` API endpoint may use the classic / legacy builder
- Windows native daemons do not yet suport BuildKit, so when building Windows images, the classic / legacy builder is used.

**Reply by thaJeztah** [reply]
Oh, and for completeness; when invoking `docker build` but with a custom builder (e.g. using `docker buildx create` to create a containerised builder or a remote builder), that builder may be used instead of the built-in BuildKit builder.

**Reply by dcermak** [reply]
Ok, but if I do not have the buildx plugin installed, then those code paths will not be reachable, despite buildkit being vendored into moby/moby?

**Reply by thaJeztah** [reply]
If you're running a current version of the CLI (older versions had a buildkit client in the CLI itself), the `buildx` component should be installed by default. It it's missing, it would use the classic builder, and print a deprecation warning in that case;

```bash
docker build .
DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
            Install the buildx component to build images with BuildKit:
            https://docs.docker.com/go/buildx/

Sending build context to Docker daemon  2.048kB
Step 1/2 : FROM busybox
...
```
